Code/src/main/java/nl/utwente/group10/ui/components/blocks/function/FunctionBlock.java

```diff
@@ -0,0 +1,259 @@
+        this.setMaxSize(Region.USE_PREF_SIZE, Region.USE_PREF_SIZE);
+        functionInfo.setMinWidth(Region.USE_PREF_SIZE);
+        functionInfo.setMaxWidth(Region.USE_PREF_SIZE);
+        invalidateConnectionState();
+    }
+    
+    /** Returns the name property of this FunctionBlock. */
+    public final String getName() {
+        return name.get();
+    }
+
+    /** Sets the name property of this FunctionBlock. */
+    public void setName(String name) {
+        this.name.set(name);
+    }
+    
+    /** Returns the StringProperty for the name of the function. */
+    public final StringProperty nameProperty() {
+        return name;
+    }
+
+    /** Returns the knot index of this FunctionBlock. */
+    public final Integer getKnotIndex() {
+        return argumentSpace.knotIndexProperty().get();
+    }
+
+    /**
+     * @param index The new knot index for this FunctionBlock.
+     */
+    public final void setBowtieIndex(int index) {
+        if (index >= -1 && index <= getAllInputs().size()) {
+            argumentSpace.setKnotIndex(index);
+        } else {
+            throw new IndexOutOfBoundsException();
+        }
+    }
+
+    @Override
+    public final List<InputAnchor> getAllInputs() {
+        return argumentSpace.getInputArguments().stream().map(a -> a.getInputAnchor()).collect(Collectors.toList());
+    }
+
+    /**
+     * @return Only the active (as specified by the knot index) inputs.
+     */
+    @Override
+    public List<InputAnchor> getActiveInputs() {
+        return getAllInputs().subList(0, getKnotIndex());
+    }
+
+    /**
+     * @param index Index of the InputAnchor to return.
+     * @return InputAnchor with the given index.
+     */
+    public InputAnchor getInput(int index) {
+        return argumentSpace.getInputArgument(index).getInputAnchor();
+    }
+
+    @Override
+    public OutputAnchor getOutputAnchor() {
+        return output;
+    }
+
+    /**
+     * @return The function signature as specified in the catalog.
+     */
+    public Type getFunctionSignature() {
+        return getPane().getEnvInstance().getFreshExprType(this.getName()).get();
+    }
+
+    @Override
+    public Type getInputSignature(int index) {
+        return getInput(index).getSignature();
+    }
+    
+    @Override
+    public Type getInputType(int index) {
+        return getInput(index).getType();
+    }
+    
+    @Override
+    public Type getOutputSignature() {
+        Type type = getFunctionSignature();
+        for (int i = 0; i < getKnotIndex(); i++) {
+            if (type instanceof FuncT) {
+                type = ((FuncT) type).getArgs()[1];
+            } else {
+                throw new FunctionDefinitionException();
+            }
+        }
+        return type;
+    }
+
+    @Override
+    public Type getOutputType() {
+        try {
+            return asExpr().analyze(getPane().getEnvInstance()).prune();
+        } catch (HaskellException e) {
+            return getOutputSignature();
+        }
```

Comment:
That would be recommended since the specialized exceptions contain a lot of useful information.